File: src/389/cpp/3p.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
 public:
  static int minimumDeletions(string &word, int k) {
	int cnt[26]{};
	for (char c : word) {
	  cnt[c - 'a']++;
	}

	ranges::sort(cnt);

	int max_save = 0;
	for (int i = 0; i < 26; ++i) {
	  int sum = 0;

	  for (int j = i; j < 26; ++j) {
		sum += min(cnt[j], cnt[i] + k); // 至多保留 cnt[i]+k 个
	  }

	  max_save = max(max_save, sum);
	}

	return word.length() - max_save;
  }
};

File: src/389/java/3p.java
import java.util.Arrays;

public class Solution {
    public static int minimumDeletions(String word, int k) {
        int[] cnt = new int[26];
        for (char c : word.toCharArray()) {
            cnt[c - 'a']++;
        }
        Arrays.sort(cnt);

        int maxSave = 0;
        for (int i = 0; i < 26; i++) {
            int sum = 0;

            for (int j = i; j < 26; j++) {
                sum += Math.min(cnt[j], cnt[i] + k);
            }

            maxSave = Math.max(maxSave, sum);
        }

        return word.length() - maxSave;
    }
}